### ISSUE TITLE
Session Start scan aborts on an empty log under zsh, so a fresh install's first session gets an error where the answer is "no observations"

### ISSUE THREAD

**Issue by theocannac** [rebelytics/one-skill-to-rule-them-all#86]
On a fresh v3 install on macOS, the first session's scan step terminated with status 1 and never entered the loop body. The cause is zsh's `nomatch`, which is on by default: an unmatched glob is an error and the command is not run at all. `for f in "$d"/*.md` therefore aborts before the first iteration, and the `[ -e "$f" ] || continue` line — which exists precisely to handle the literal-glob case — is dead code in zsh. The block's own integrity check (`n > 0 && parsed == 0`) never runs either, so the one guard meant to distinguish "the log is empty" from "the scan command is broken" is itself absent exactly where that distinction matters.

The failure is confined to the empty-log case; with at least one `.md` present both snippets behave correctly under zsh. That confines it to the first session of every new install — the session where the agent has the least evidence for telling an empty log from a broken command, and the one where an error at step 2 is most likely to be read as "storage setup failed". The id snippet at line 298 has a milder version: `[ -n "$(ls "$d"/*.md 2>/dev/null)" ]` writes `no matches found` to stderr on every fresh log, though the guard's outcome stays correct because the substitution simply comes back empty.

Reproduced with the literal strings from SKILL.md in a clean shell (`zsh -f -c`, no rc files), on macOS 15 / zsh 5.9. Empty log: the scan loop exits 1 with `zsh:2: no matches found: skill-observations/observation-log/*.md` and prints nothing. Same input under `sh`: the loop completes normally. Non-empty log: both shells fine. The same snippet appears in `references/observation-log.md`, and a variant of the count in `references/migration.md`, so a fix needs all of them.

On the relationship to #77: the three modes catalogued there are the snippet erroring out on the wrong flavour of a tool, succeeding but returning nothing, and never parsing. This is a fourth — the snippet parses, every tool is the right one, POSIX semantics are correct, and it still fails on the default interactive shell of the platform the skill's own compatibility list names first. Portability review that stops at "is it POSIX?" cannot catch it, because the divergence is a shell option, not a dialect.

Suggested fix, if useful: replace the glob-driven iteration and count with `find`, which returns empty without erroring in every shell. Given that the skill controls the filename scheme (`NNNN-slug.md`, kebab-case, no whitespace), `for f in $(find "$d" -maxdepth 1 -name '*.md')` is safe here and keeps the `parsed` counter in the current shell — a `while read` pipeline would put it in a subshell and silently lose it, which would trade one silent failure for another. The count line becomes `n=$(find "$d" -maxdepth 1 -name '*.md' | wc -l | tr -d ' ')`.

**Comment by rlucho** [comment]
The parenthetical in the suggested fix — *"a `while read` pipeline would put it in a subshell and silently lose it, which would trade one silent failure for another"* — is correct, and I hit that exact failure from a different direction. Adding the measurement in case it strengthens the argument for keeping `parsed` out of any pipeline by construction rather than by author discipline.

**What happened.** Windows/Git-Bash, v3.0.1. The obvious way to keep the scan's output manageable on a large log is to filter it: `done | grep -E '^(id|title|status):'`. That moves the whole `for` loop into a subshell, so every `parsed=$(( parsed + 1 ))` is discarded and the parent reads 0. The run printed **13 well-formed frontmatter blocks and then `SCAN COMMAND BROKEN — 13 files present, 0 headers parsed`**.

**Why that is worse than an ordinary bug.** The guard's false positive is textually identical to its true positive — both say "files exist, nothing parsed". At the point of reading the output there is nothing to distinguish "the log is corrupt" from "the guard's own state was relocated", and the correct verdict had to be recovered by noticing that the headers were sitting right above the alarm. A guard whose failure mode is indistinguishable from the condition it guards against is worse than no guard, because it spends the reader's trust on both sides.

**What makes it reachable.** The edit that breaks it is adding a filter after `done` — that reads as a *formatting* change, not a logic change, so nothing prompts a re-check. `while read` is the version an author might think twice about; a `grep` on the end is not. So the hazard is not really about which loop form is chosen, it is that the guard's state lives somewhere an innocuous edit can move.

**Suggested strengthening of the fix already proposed here.** Beyond switching to `find` (which I agree with, and which fixes the `nomatch` abort), derive `parsed` from the *emitted output* rather than from a counter the loop maintains — capture once, then count:

```sh
out=$(for f in $(find "$d" -maxdepth 1 -name '*.md'); do
        awk 'NR==1 && /^---[[:space:]]*$/ {fm=1; next}
             fm && /^---[[:space:]]*$/ {exit}
             fm' "$f"
        printf '%s\n' '---'
      done)
n=$(find "$d" -maxdepth 1 -name '*.md' | wc -l | tr -d ' ')
parsed=$(printf '%s\n' "$out" | grep -c '^id:')
printf '%s\n' "$out"
```

`parsed` is then a property of the artefact the command actually produced, so no later reshaping of the output can silently invalidate it, and adding a filter to the final `printf` is harmless. The general form, if it is worth a line next to the snippet: **prefer guards derived from what a command produced over guards derived from bookkeeping it kept while producing it.**

(Cross-referencing #87 — that block's trailing `&&` also leaks exit 1 on the success path on Git-Bash, so on a harness that surfaces status the same step can report failure three different ways for three unrelated reasons.)

**Comment by rebelytics** [comment]
Fixed in `main` and shipped in v3.1.0 today — the scan enumerates with find; verified on an empty log in bash and sh (no zsh in my sandbox — if you can confirm on zsh that would close the loop). Thanks for the report.
